synapse/http/federation/srv_resolver.py

```diff
@@ -83,6 +83,35 @@ def pick_server_from_list(server_list):
     raise RuntimeError("pick_server_from_list got to end of eligible server list.")
 
 
+def _sort_server_list(server_list):
+    """Given a list of SRV records sort them into priority order and shuffle
+    each priority with the given weight.
+    """
+    priority_map = {}
+
+    for server in server_list:
+        priority_map.setdefault(server.priority, []).append(server)
+
+    results = []
+    for priority in sorted(priority_map):
+        servers = priority_map.pop(priority)
+
+        while servers:
+            total_weight = sum(s.weight for s in servers)
+            target_weight = random.randint(0, total_weight)
```

Comment:
Turns out this algorithm is the same as the one in the [RFC2782](https://tools.ietf.org/html/rfc2782), so I've documented it as such.

Reply:
I think the off-by-one error is still present?

Reply:
Unfortunately its complicated to change the algorithm to work with zero weights, as eventually you end up trying to do `randint(1, 0)` and everything blows up. I don't know whether we want to use a different algorithm which is correctly weighted, or just accept the off by one.

Reply:
wait, the RFC2782 algorithm has the off-by-one error? that's... bonkers.

To be clear: if you have two servers, each with a weight of 1, we'll currently use one twice as often as the other.

can we just special-case `total_weight == 0`? (always return the first server, or something)

Reply:
Quite. So the RFC says that its valid to have multiple servers that have a weight of zero, which means:

> In the presence of records containing weights greater than 0, records with weight 0 should have a very small chance of being selected.

We could just always order weights of zero last, and shuffle them randomly if there are multiple of them? That I think should work, though ignores that section of RFC a bit

Reply:
> We could just always order weights of zero last, and shuffle them randomly if there are multiple of them? That I think should work, though ignores that section of RFC a bit

it doesn't matter if we put them at the start or the end: a server with a weight of 0 has a 0% chance of being used in a sane algorithm. Perhaps we should do `randint(0, total_weight)` if there are any servers with a zero weight, and `randint(1, total_weight)` otherwise?

I guess it doesn't really matter, since I doubt anyone is really using SRV records to load-balance servers...

Reply:
The idea is that synapse will sequentially try each server on the list until it finds one it can connect to, so putting the zero weights at the end of the returned list means that they will get tried last (modulo priority).

Reply:
ohhh, right. Fair enough. sgtm then.